panos_upgrade_assurance/check_firewall.py

```diff
@@ -870,6 +864,185 @@ def check_ssl_cert_requirements(self, rsa: dict = {}, ecdsa: dict = {}) -> Check
         result.status = CheckStatus.SUCCESS
         return result
 
+    def check_non_finished_jobs(self) -> CheckResult:
+        """Check for any job with status different than FIN.
+
+        # Returns
+
+        CheckResult: Object of [`CheckResult`](/panos/docs/panos-upgrade-assurance/api/utils#class-checkresult) class taking \
+            value of:
+
+        * [`CheckStatus.SUCCESS`](/panos/docs/panos-upgrade-assurance/api/utils#class-checkstatus) when all jobs are in FIN state.
+        * [`CheckStatus.FAIL`](/panos/docs/panos-upgrade-assurance/api/utils#class-checkstatus) otherwise, `CheckResult.reason`
+            field contains information about the 1<sup>st</sup> job found with status different than FIN (job ID and the actual
+            status).
+
+        """
+        result = CheckResult()
+
+        all_jobs = self._node.get_jobs()
+
+        for jid, job in all_jobs.items():
+            if job["status"] != "FIN":
+                result.reason = f"At least one job (ID={jid}) is not in finished state (state={job['status']})."
+                return result
+
+        result.status = CheckStatus.SUCCESS
+        return result
+
+    def _calculate_time_distance(self, schedule_type: str, schedule: dict) -> (int, str):
+        """A method that calculates the time distance between two `datetime` objects.
+
+        :::note
+        This method is used only by [`CheckFirewall.check_scheduled_updates()`](/panos/docs/panos-upgrade-assurance/api/check_firewall#checkfirewallcheck_scheduled_updates) method and it expects some information
+        to be already available.
+        :::
+
+        # Parameters
+
+        schedule_type (str): A schedule type returned by PanOS, can be one of: `every-*`, `hourly`, `daily`, `weekly`,
+            `real-time`.
+        schedule (dict): Value of the `recurring` key in the API response, see [`FirewallProxy.get_update_schedules()`](/panos/docs/panos-upgrade-assurance/api/firewall_proxy#firewallproxyget_update_schedules)
+            documentation for details.
+
+        # Raises
+
+        MalformedResponseException: Thrown then the `schedule_type` is not recognizable.
+
+        # Returns
+
+        tuple(int, str): A tuple containing the calculated time difference (in minutes) and human-readable description.
+
+        """
+        time_distance = 0
+        details = "unsupported schedule type"
+
+        if schedule_type == "daily":
+            occurrence = schedule["at"]
+            next_occurrence = datetime.strptime(f"{str(self._mp_now.date())} {occurrence}", "%Y-%m-%d %H:%M")
+
+            if self._mp_now > next_occurrence:
+                next_occurrence = next_occurrence + timedelta(days=1)
+            diff = next_occurrence - self._mp_now
+            time_distance = floor(diff.total_seconds() / 60)
+            details = f"at {next_occurrence.time()}"
+
+        elif schedule_type == "hourly":
+            time_distance = 60
```

Comment:
why don't we calculate the time difference when it's hourly?

Reply:
check [this comment](https://github.com/PaloAltoNetworks/pan-os-upgrade-assurance/pull/106#discussion_r1289814352)

Reply:
I see but I think we can still calculate it since this is a time difference function, and `check_scheduled_updates` would still work the same. This is also the current case with the "daily" schedule when there might be less than 60 minutes for the next occurrence we return the exact difference.

Reply:
if we would calculate the exact difference, the only thing that would change would be the way we present the next occurrence in the comments.

Reply:
I understand but it looks inconsistent with the "daily" schedule where we provide the exact minutes of difference if it's less than 60 minutes, isn't it?